nixos/modules/services/networking/spacecookie.nix

```diff
@@ -0,0 +1,87 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.spacecookie;
+  defaultUser = "spacecookie";
+  configFile = pkgs.writeText "spacecookie.json" (lib.generators.toJSON {} {
+    hostname = cfg.hostname;
+    port = cfg.port;
+    user = cfg.user;
+    root = cfg.root;
+  });
+in {
+
+  options = {
+
+    services.spacecookie = {
+
+      enable = mkEnableOption "spacecookie";
+
+      hostname = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = "The hostname the service is reachable via. Clients will use this hostname for further requests after loading the initial gopher menu.";
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 70;
+        description = "Port the gopher service should be exposed on.";
+      };
+
+      root = mkOption {
+        type = types.path;
+        default = "/srv/gopher";
+        description = "The root directory spacecookie serves via gopher.";
+      };
+
+      user = mkOption {
+        type = types.str;
+        default = "spacecookie";
+        description = "User the daemon should be run as.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.sockets.spacecookie = {
+      description = "Socket for the Spacecookie Gopher Server";
+      wantedBy = [ "sockets.target" ];
+      listenStreams = [ "[::]:${toString cfg.port}" ];
+      socketConfig = {
+        BindIPv6Only = "both";
+        SocketUser = cfg.user;
+        SocketGroup = cfg.user;
+        SocketMode = 0660;
+      };
+    };
+
+    systemd.services.spacecookie = {
+      description = "Spacecookie Gopher Server";
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "spacecookie.socket" ];
+
+      serviceConfig = {
+        Type = "notify";
+        ExecStart = "${pkgs.haskellPackages.spacecookie}/bin/spacecookie ${configFile}";
+        FileDescriptorStoreMax = 1;
+        StandardError = "journal";
+        NotifyAccess = "main";
+        User = cfg.user;
+      };
+    };
+
+    users.users = mkIf (cfg.user == defaultUser) {
+      ${defaultUser} = {
+        group = defaultUser;
+      };
+    };
+
+    users.groups = mkIf (cfg.user == defaultUser) {
+      ${defaultUser} = {};
+    };
+  };
```

Comment:
What kind of state does the application persist in general?

If it persists nothing, or could be restricted to `/var/lib/spacecookie/…`, it might be perfectly fine to just let it run with `DynamicUser=true`, and not worry about setting up users and groups manually at all.

Reply:
Good suggestions, this should work fine, as it doesn't depend on any state except for the saved directory and the config file.

Reply:
This should work fine, most of the `ProtectSystem` stuff is for `/usr/` and `/etc` really…

PTAL at https://www.freedesktop.org/software/systemd/man/systemd.exec.html#Sandboxing.